core/field_number.js

```diff
@@ -119,9 +119,8 @@ Blockly.FieldNumber.prototype.doClassValidation_ = function(opt_newValue) {
   newValue = newValue.replace(/O/ig, '0');
   // Strip out thousands separators.
   newValue = newValue.replace(/,/g, '');
-
-  // Clean up number.
-  var n = Number(newValue || 0);
+  var hex = newValue.match(/^0xX+$/g);
+  var n = hex ? parseInt(newValue, 16) : Number(newValue || 0);
```

Comment:
This also isn't necessary, Number('0xabc') and parseInt('0xabc', 16) both return the same value, 2748. Converting a 3 letter hex color to a 6 letter number is a bit trickier than this.